addons/stock/models/stock_warehouse.py

```diff
@@ -395,6 +394,14 @@ def _create_or_update_route(self):
             rules_list = self._get_rule_values(
                 rules, values=route_data['rules_values'])
             # Create/Active rules
+            if route_data.get('last_rule_cancel') and rules_list and rules_list[-1].get('propagate_cancel'):
```

Comment:
i would try to move this logic into _get_rule_values
furthermore there's already a variable named "first_rule" there :)
use enumerate on route_values to get the index